File: 1489-find-critical-and-pseudo-critical-edges-in-minimum-spanning-tree/1489-find-critical-and-pseudo-critical-edges-in-minimum-spanning-tree.py
class UnionFind:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
        self.rank = [1 for _ in range(n)]

    def find(self, node):
        if self.parent[node] == node:
            return node
        self.parent[node] = self.find(self.parent[node])
        return self.parent[node]

    def unionNode(self, n1, n2):
        p1 = self.find(n1)
        p2 = self.find(n2)

        if p1 == p2:
            return False

        r1 = self.rank[p1]
        r2 = self.rank[p2]

        if r1 >= r2:
            self.parent[p2] = p1
            self.rank[p1] += self.rank[p2]
        else:
            self.parent[p1] = p2
            self.rank[p2] += self.rank[p1]

        return True

        

class Solution:
    def findCriticalAndPseudoCriticalEdges(self, n: int, edges: List[List[int]]) -> List[List[int]]:
        # append idx to edges
        for i, e in enumerate(edges):
            e.append(i)

        # sort edges
        edges.sort(key = lambda e : e[2])

        # find minWeight with MST
        minWeight = self.getWeight(n, edges)

        # loop through edges, try remove one. If weight > minWeight, it's critical
        # force adding edge, if weight == minWeight, it's pseudo

        critical = []
        pseudo = []

        for n1, n2, testW, i in edges:
            uf = UnionFind(n)
            weight = 0

            # try remove
            for start, end, w, j, in edges:
                if i == j:
                    continue

                if (uf.unionNode(start, end)):
                    weight += w

            if max(uf.rank) != n or weight > minWeight: # Got smaller weight or can't get MST at all
                critical.append(i)
                continue

            # try force adding
            uf = UnionFind(n)
            uf.unionNode(n1, n2)
            weight = testW

            # try remove
            for start, end, w, idx, in edges:
                if (uf.unionNode(start, end)):
                    weight += w

            if weight == minWeight:
                pseudo.append(i)

        return [critical, pseudo]


    def getWeight(self, n, edges):
        uf = UnionFind(n)
        res = 0

        for start, end, weight, idx, in edges:
            if (uf.unionNode(start, end)):
                res += weight

        return res

File: 1489-find-critical-and-pseudo-critical-edges-in-minimum-spanning-tree/1489-find-critical-and-pseudo-critical-edges-in-minimum-spanning-tree.cpp
class UnionFind {
public:
    vector<int> parent;
    vector<int> rank;

    UnionFind(int n) : parent(n), rank(n, 1) {
        for (int i = 0; i < n; i++) parent[i] = i;
    }

    int find(int node) {
        if (parent[node] == node) return node;
        return parent[node] = find(parent[node]);
    }

    bool unionNode(int n1, int n2) {
        int p1 = find(n1);
        int p2 = find(n2);

        if (p1 == p2) return false;

        if (rank[p1] >= rank[p2]) {
            parent[p2] = p1;
            rank[p1] += rank[p2];
        } else {
            parent[p1] = p2;
            rank[p2] += rank[p1];
        }
        return true;
    }
};

class Solution {
public:
    vector<vector<int>> findCriticalAndPseudoCriticalEdges(int n, vector<vector<int>>& edges) {
        // append idx to edges
        for (int i = 0; i < edges.size(); i++) {
            edges[i].push_back(i);
        }

        // sort edges
        sort(edges.begin(), edges.end(), [](auto& a, auto& b) {
            return a[2] < b[2];
        });

        // find minWeight with MST
        int minWeight = getWeight(n, edges);

        // loop through edges, try remove one. If weight > minWeight, it's critical
        // force adding edge, if weight == minWeight, it's pseudo
        vector<int> critical, pseudo;

        for (auto& e : edges) {
            int n1 = e[0], n2 = e[1], testW = e[2], i = e[3];

            // try remove
            UnionFind uf(n);
            int weight = 0;
            for (auto& other : edges) {
                if (other[3] == i) continue;
                if (uf.unionNode(other[0], other[1])) {
                    weight += other[2];
                }
            }

            // Got smaller weight or can't get MST at all
            if (*max_element(uf.rank.begin(), uf.rank.end()) != n || weight > minWeight) {
                critical.push_back(i);
                continue;
            }

            // try force adding
            UnionFind uf2(n);
            uf2.unionNode(n1, n2);
            weight = testW;

            for (auto& other : edges) {
                if (uf2.unionNode(other[0], other[1])) {
                    weight += other[2];
                }
            }

            if (weight == minWeight) {
                pseudo.push_back(i);
            }
        }

        return {critical, pseudo};
    }

    int getWeight(int n, vector<vector<int>>& edges) {
        UnionFind uf(n);
        int res = 0;

        for (auto& e : edges) {
            if (uf.unionNode(e[0], e[1])) {
                res += e[2];
            }
        }

        return res;
    }
};